### DISCUSSION TITLE
Need help understanding profile stuff

### DISCUSSION THREAD

**Discussion by speedgeil** [ImageMagick/ImageMagick#5151] [Digital Image Processing]
Hello,

As far as i understand an ICC profile will not change pixel data. So when i would apply an ICC profile at the beginning of my IM command and later add another one, the first profile would be exchanged with the second profile?

Does a embedded profile need to be stripped before i would apply a different profile or would i just "overwrite" it?

What if i wanted to misuse an ICC profile as a Lut and later convert to sRGB? Would it be possible to change pixel data with the ICC profile before transforming to sRGB?

Sorry, i'm a bit confused by what i've read in the docs and what i read elsewhere.

**Comment by snibgo** [comment]
An ICC profile is metadata that defines the colorimetric properties of an image. So an image with an embedded ICC profile should look the same on any computer, with any software, provided the software is "colour managed", and within the limitations of the hardware.

We can do three operations on embedded profiles:

1. We can *remove* the profile. This changes metadata only. In IM, "+profile ICC" will remove an embedded ICC profile, if it has one. We can also "-strip", which will remove any ICC profile, as well as other profiles and other metadata.

2. If an image has no embedded image, we can *assign* one. This changes metadata only, embedding the profile. In IM, "+profile abc.icc" does this.

3. If an image has an embedded profile, we can *convert* the image from that profile to a new one. This changes pixel values and metadata, removing the old profile and embedding the new one. In IM, "+profile abc.icc" does this.

In IM and most other software, an image with no embedded profile is assumed to be encoded as sRGB.

> So when i would apply an ICC...

ICC profiles are not "applied" (or "added", as some people say). A profile is *assigned* or *removed*, or an image is *converted* from one profile to another.

If a command has "-profile" twice or more on an image, the first "-profile" will either assign a profile (if one was not already embedded), or convert to a new profile (if one was already embedded). The second (and third and fourth) "-profile" will convert from the current profile to the new profile.

> Does a embedded profile need to be stripped before i would apply a different profile or would i just "overwrite" it?

If you want to *convert* to a new profile, then stripping the old one first won't work, because IM won't know how to convert from the old profile to the new one.

> What if i wanted to misuse an ICC profile as a Lut and later convert to sRGB? Would it be possible to change pixel data with the ICC profile before transforming to sRGB?

Yes, you can do that. A single profile can't be used as a LUT. Remember, the way to change pixels with profiles is to convert from one profile to a different profile.

**Comment by speedgeil** [comment]
Thank you very much for the detailed explanation. There is one thing i still don't understand.

What if i want to "burn in destructively" ICC properties into pixel values?
Somehow i only manage to asssign one ICC profile and then assign the color space. But i'm not able to assign two profiles and convert it to a color space (like AdobeRGB for example).
I understood i need to use +profile xxx.icc and i tried it in various combinations but had no luck.
What i basically want to do is using ICC profiles like luts on top of each other and convert to AdobeRGB:

input.dng -strip -profile first.icc  -profile second.icc -profile AdobeRGB1998.icc output.dng

I read, -strip could be a problem in this combination but well, i  just need to strip all meta.

*Edit* i now see that it works well with ICC profiles i made in PS and other editing software. But it doesn't with device specific profiles, like IT-8 profiles i made with dcamprof or BasicColor. StdOut gives me a "unable to create color transform `icc' @ error/profile.c/ProfileImage/1329". Any clue ?

**Comment by snibgo** [comment]
> What if i want to "burn in destructively" ICC properties into pixel values?

Sorry, I don't understand what you want.

> But i'm not able to assign two profiles and convert it to a color space (like AdobeRGB for example).

What would it mean to "assign two profiles"? Only one profile is needed to define the colorimetric properties of an image. What would the second profile do? What do you want it to do?

> I understood i need to use +profile xxx.icc ...

"+profile" will remove a profile, as I mention above.

> input.dng -strip -profile first.icc -profile second.icc -profile AdobeRGB1998.icc output.dng

This will read input.png, then "-strip" will remove any ICC colour profiles (if any are present). Then "-profile first.icc" will *assign* first.icc. Then "-profile second.icc" will *convert* from colorspace "first" to colorspace "second". Then "-profile AdobeRGB1998.icc" will convert from colorspace "second" to colorspace "AdobeRGB1998".

So "-profile second.icc" is redundant. If you remove it, the command should have the same effect (aside from issues with rounding, and clipped out-of-gamut pixels).

> i now see that it works well with ICC profiles i made in PS and other editing software. But it doesn't with device specific profiles, like IT-8 profiles i made with dcamprof or BasicColor. Any clue ?

My comments should apply to any ICC colour profile, however that profile was made. I'm not familiar with dcamprof or BasicColor. Please link to one of more such profiles, and your command, and any required input files so I can try to reproduce the problem.

**Comment by speedgeil** [comment]
I will try to explain what i'm up to. maybe that makes it understandable.

I want to assign an IT8 profile before calculating things like autogamma etc. Hence i want to change pixel values with the IT8 device link profile. And then i want to the image to have a special look with a second ICC profile. Maybe i want to change some hues and leave saturation unchanged for example. And the resulting color space should be Adobe RGB. Like working with 3D luts as adjustment layers in PS.

The example pseudo code was just for explaining the processing chain ( the order in which the profiles should apply). 

BTW the basiccolor/dcamprof device link profiles will work fine as long i won't try to transform to a different profile afterwards. Then i will get the stdout i mentioned before.

**Comment by snibgo** [comment]
> And then i want to the image to have a special look with a second ICC profile.

*Converting* an image from one profile to another won't change colours (aside from rounding and gamut issues). It changes pixel values, but doesn't change colours.

*Assigning* a colour profile won't change pixel values, but it may change colours. (If the assigned profile is factually incorrect).

Similarly, *removing* a profile won't change pixel values, but will change colours unless the colorspace was sRGB, which is the default.

It may help to imagine that ICC colour profiles are like units of length.

If I say, "The distance from my house to the shop is 12.3", it doesn't mean anything because there are no units.

If I say, "The distance from my house to the shop is 12.3 miles", it does mean something. Now we can *convert* that distance to a different unit.

"The distance from my house to the shop is 19.8 kilometres." This is the same length. The number is different, but the length is the same (within rounding error).

And we can *convert* that to another unit:

"The distance from my house to the shop is 19800 metres."

At any stage, we may remove the units:

"The distance from my house to the shop is 19800."

This may be useful, provided there is an overall rule like "measurements with no units are assumed to be in metres", equivalent to "images without embedded profiles are assumed to be sRGB".

We can also *assign* a unit:

"The distance from my house to the shop is 19800 miles."

This is not useful, except where we have some other knowledge about the actual units.

**Comment by speedgeil** [comment]
Thank you so much for your patience. I imagine it's sometimes annoying. I saw you were answering similar questions in other threads over and over again.

So yes, i think then it's the same when opening an image in Photoshop without color profile. Then i can assign a profile of my choice. And when the image is rendered i can choose between assigning or converting to a different color space.

So probably i have a misunderstanding somewhere because i'm used to working with luts in PS (as adjustment layers). And when working with layers we can do whatever we want. Opacity, blend mode and whatnot.

So i thought about another example. Let's say i have scanned a negative and i made an it8 icc device link profile. Now i can assign the it8 profile to the negative scan in IM and then do some calculations on the image . One of them may be inverting. Wouldn't the IT8 profile as metadata stay uninverted while the "underlying" pixel data is inverted? And in return the result would be wrong?
This is just hypothetical, as it8 profiling for negative film wouldn't make sense...
However, i think before inverting i may rather somehow change colors in the pixel data for having an exact result after inversion. Is something like that possible in IM? From what i understood from your explanations i can't do that. Just to make sure.

Thank you again.

**Reply by fmw42** [reply]
Why do you need profiles for image processing tasks. You can use -clut or -hald-lut or many other processing methods. You shouldn't need a profile to simply assign that processing.

**Comment by snibgo** [comment]
> Let's say i have scanned a negative and i made an it8 icc device link profile.

"IT8" is a type of colour chart.

I am mostly familiar with ICC *device* profiles (just "device", not "device link"). These define colour characteristics of an image in terms of a standard PCS (profile connection space). The PCS is usually XYZ or Lab. Other PCSs are possible, but I have never seen any. 

So AdobeRGB1988.icc, for example, defines the transformation in both directions between AdobeRGB1988 colorspace and XYZ (or Lab). So if an image has a different embedded ICC profile, and we want to convert the image from that to AdobeRGB, software can use the embedded profile to convert the image from that colorspace to a PCS, then use the new profile to convert from the PCS to AdobeRGB colorspace.

Hence we need one ICC device profile per colorspace. If we have 20 colorspaces, we need 20 profiles.

An ICC device link is different. It combines the effect of two device profiles, so it defines the transformation from one colorspace to another. For example, it might define the transformation between AdobeRGB and a CMYK SWOP colorspace.

When using device link profiles we need need one profile for every conversion that we might want. If we have 20 colorspaces, we need 400 profiles. But they should be faster to use than ordinary device profiles, and are useful when input and output devices both have more than three channels (eg CMYK).

I have no experience with *device link* profiles. I have never seen an image with an embedded device link  profile. This is not surprising, as they logically don't need to be embedded; only the second half of the profile, that functions as a simple *device* profile needs to be embedded.

IM uses LCMS for colour management. I don't know if LCMS can perform conversions using *device link* profiles. If it can, I don't know if IM has implemented this. Perhaps this explains your error message.

For more on the types of ICC colour profiles, see https://www.color.org/specification/ICC.1-2022-05.pdf

On inverting a scanned negative: remember that an embedded ICC device profile defines the colorimetric properties of an image. If you negate the image, the embedded profile is then wrong; it *doesn't* define the colours.

Ideally, what you want is a photograph of an IT8 target taken on a camera containing the film you are scanning. This will be a negative. From that negative, make a device profile.

Then scan your negatives. For each, *assign* the profile from the previous paragraph, then *convert* to whatever profile you want: sRGB, AdobeRGB, Rec2020 or whatever. This conversion will negate the negative image, giving a positive image.

**Comment by speedgeil** [comment]
Thank you for the detailed explanation. Indeed this makes a lot of sense.
What i had in mind is Vuescan for example. It allows profiling the scanner by scanning an IT8 target. I guess it profiles the sensor response. It also offers profiling a film stock with the same target. That is then slides of course. But it will use the scanner icc profile for scanning negatives as well

**Comment by snibgo** [comment]
Ideally, you need to calibrate (ie make an ICC device profile) for the combination of your scanner and the film you will be scanning. For example, see https://www.hamrick.com/vuescan/html/vuesc19.htm:

> Film profiling is the process of determining the precise color characteristics of color negative film. To do this, you need to take a picture of an IT8 target and scan this frame. Make sure the picture is rectangular on the film.
